stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2005 @@
+        let origin_nonce = transaction.get_origin_nonce();
+        let Some(origin_signer_id) = signer_slot_ids.get(&origin_address) else {
+            debug!(
+                "Signer #{}: Unrecognized origin address ({origin_address}). Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        };
+        let Ok(account_nonce) = retry_with_exponential_backoff(|| {
+            stacks_client
+                .get_account_nonce(&origin_address)
+                .map_err(backoff::Error::transient)
+        }) else {
+            warn!(
+                "Signer #{}: Unable to get account for transaction origin address: {origin_address}. Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        };
+        // TODO: add a check that we don't have two conflicting transactions in the same block from the same signer. This is a potential attack vector (will result in an invalid block)
+        if origin_nonce < account_nonce {
+            debug!("Signer #{}: Received a transaction with an outdated nonce ({account_nonce} < {origin_nonce}). Filtering ({}).", self.signer_id, transaction.txid());
+            return None;
+        }
+        if transaction.is_mainnet() != self.mainnet {
+            debug!(
+                "Signer #{}: Received a transaction with an unexpected network. Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        }
+        let Ok(valid) = retry_with_exponential_backoff(|| {
+            self.verify_payload(
+                stacks_client,
+                &transaction,
+                *origin_signer_id,
+                current_reward_cycle,
+            )
+            .map_err(backoff::Error::transient)
+        }) else {
+            warn!(
+                "Signer #{}: Unable to validate transaction payload. Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        };
+        if !valid {
+            debug!(
+                "Signer #{}: Received a transaction with an invalid payload. Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        }
+        debug!(
+            "Signer #{}: Expect transaction {} ({transaction:?})",
+            self.signer_id,
+            transaction.txid()
+        );
+        Some(transaction)
+    }
+
+    ///Helper function to verify the payload contents of a transaction are as expected
+    fn verify_payload(
+        &self,
+        stacks_client: &StacksClient,
+        transaction: &StacksTransaction,
+        origin_signer_id: u32,
+        current_reward_cycle: u64,
+    ) -> Result<bool, ClientError> {
+        let TransactionPayload::ContractCall(payload) = &transaction.payload else {
```

Comment:
On some level, I'm curious why we even need to validate the structure of the signer-submitted transaction?  Do we have a problem with signers submitting zero-fee non-voting transactions?  Because, that sounds more like a feature to me -- it gives people yet another reason to want to be a stacker.  For example, a stacker could submit oracle price feeds into the Stacks blockchain this way (something that Pyth is struggling to do right now, due to how expensive they are).